### PR TITLE
[EIP1014]: eip1014 support

### LINKED ISSUE #2869: Add address computation for deterministic EVM smart contracts
**Is your feature request related to a problem? Please describe.**

Account abstraction [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) is powered by deterministic smart contracts created via CREATE2 opcode ([EIP-1014](https://eips.ethereum.org/EIPS/eip-1014)). This is done in order to not make the contract deployment necessary requirement for receiving funds on SCWs. The contract will be deployed as soon as the first transaction is initiated from the wallet, however users should still be able to receive funds from the very beginning, without deploying the smart contract.

For wallet-core to support account abstraction and ERC-4337, it's important to have the same level of API coverage as for EOAs. And calculating the address is the first and a very basic capability that has to be supported.

**Describe the solution you'd like**

**Public API**

Let's go over the steps to derive ERC-4337 compatible SCW address from a mnemonic seed phrase in Swift:

1. Find out the owner address based on the private key derived from the mnemonic seed phrase
```
let ownerAddress = CoinType.Ethereum.deriveAddress(privateKey: privateKey)
```
2. Specify the factory address. Factory is a smart contract which is responsible for deploying every user's smart contract wallet. It will presumably be static, but it's still better to specify it from the client just in case. ([source code](https://github.com/eth-infinitism/account-abstraction/blob/5a1ad4072438d9e9f7c934b66464dc05a4b37d02/contracts/samples/SimpleAccountFactory.sol#L28-L38))
```
let factoryAddress = AnyAddress(string: "0xd9145CCE52D386f254917e481eB44e9943F39138", coin: .ethereum)! 
```
3. Specify the address of SCW logic implementation. It is dynamic and can be updated from clients. ([example source code](https://github.com/eth-infinitism/account-abstraction/blob/5a1ad4072438d9e9f7c934b66464dc05a4b37d02/contracts/samples/SimpleAccount.sol))
```
let implAddress = AnyAddress(string: "0x5C9eb5D6a6C2c1B3EFc52255C0b356f116f6f66D", coin: .ethereum)! 
```

4. Encode the [SimpleAccount.sol initializer](https://github.com/eth-infinitism/account-abstraction/blob/5a1ad4072438d9e9f7c934b66464dc05a4b37d02/contracts/samples/SimpleAccount.sol#L80-L82) which will later be [passed to the factory](https://github.com/eth-infinitism/account-abstraction/blob/5a1ad4072438d9e9f7c934b66464dc05a4b37d02/contracts/samples/SimpleAccountFactory.sol#L36):
```
let initializerABI = EthereumAbiFunction(name: "initialize")
function.addParamAddress(val: ownerAddress.data, isOutput: false)
let initBytecode = EthereumAbi.encode(fn: function)
```

5. And finally call the ***NEW*** API that we request in this task which will return a ERC-4337 compatible SCW address:
```
let erc4337Address = coin.deriveERC4337Address(privateKey, factoryAddress, implementationAddress, initBytecode);
```

Here is a code snippet based on the previous 5 steps with a test case that has to be passing:
```
let ownerAddress = AnyAddress(string: "0xA5a1dddEF094095AfB7b6e322dE72961DF2e1988", coin: .ethereum)! 
let factoryAddress = AnyAddress(string: "0xd9145CCE52D386f254917e481eB44e9943F39138", coin: .ethereum)!
let implAddress = AnyAddress(string: "0x5C9eb5D6a6C2c1B3EFc52255C0b356f116f6f66D", coin: .ethereum)! 
let initializerABI = EthereumAbiFunction(name: "initialize")
function.addParamAddress(val: ownerAddress.data, isOutput: false)
let initBytecode = EthereumAbi.encode(fn: function) // 0xc4d66de8000000000000000000000000a5a1dddef094095afb7b6e322de72961df2e1988
let erc4337Address = coin.deriveERC4337Address(privateKey, factoryAddress, implAddress, initBytecode);
XCTEqual(erc4337Address.string, 0xf6548aAD02E0B355a0778ACc7799FD83B2506b58) // final SCW address
```

**Private API**

Now let's try to understand the implementation of `deriveERC4337Address`. It will be using few other methods that needs to be added to Wallet Core and probably shouldn't be exposed to the public API.

(a mix of pseudocode and ChatGPT responses, as I'm not very good with c++)
```
address deriveERC4337Address(privateKey, factory_address, implementation_address, init_bytecode) {
  const proxy_contract_init_code = getProxyContractInitCode(implementation_address, init_bytecode);
  return create2Address(factory_address, 0, keccak256(proxyContractInitCode));
}

// Test case:
// implementation_address: 0x5C9eb5D6a6C2c1B3EFc52255C0b356f116f6f66D
// init_bytecode: 0xc4d66de8000000000000000000000000a5a1dddef094095afb7b6e322de72961df2e1988
// returns: 0x608060405260405162000c6638038062000c6683398181016040528101906200002991906200058b565b6200003d828260006200004560201b60201c565b5050620007e2565b62000056836200008860201b60201c565b600082511180620000645750805b156200008357620000818383620000df60201b620000371760201c565b505b505050565b62000099816200011560201b60201c565b8073ffffffffffffffffffffffffffffffffffffffff167fbc7cd75a20ee27fd9adebab32041f755214dbc6bffa90cc0225b39da2e5c2d3b60405160405180910390a250565b60606200010d838360405180606001604052806027815260200162000c3f60279139620001eb60201b60201c565b905092915050565b6200012b816200027d60201b620000641760201c565b6200016d576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401620001649062000678565b60405180910390fd5b80620001a77f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b620002a060201b620000871760201c565b60000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b60606000808573ffffffffffffffffffffffffffffffffffffffff1685604051620002179190620006e7565b600060405180830381855af49150503d806000811462000254576040519150601f19603f3d011682016040523d82523d6000602084013e62000259565b606091505b50915091506200027286838387620002aa60201b60201c565b925050509392505050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b606083156200031b576000835114156200031257620002cf856200027d60201b60201c565b62000311576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401620003089062000750565b60405180910390fd5b5b8290506200032e565b6200032d83836200033660201b60201c565b5b949350505050565b6000825111156200034a5781518083602001fd5b806040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401620003809190620007be565b60405180910390fd5b6000604051905090565b600080fd5b600080fd5b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000620003ca826200039d565b9050919050565b620003dc81620003bd565b8114620003e857600080fd5b50565b600081519050620003fc81620003d1565b92915050565b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b62000457826200040c565b810181811067ffffffffffffffff821117156200047957620004786200041d565b5b80604052505050565b60006200048e62000389565b90506200049c82826200044c565b919050565b600067ffffffffffffffff821115620004bf57620004be6200041d565b5b620004ca826200040c565b9050602081019050919050565b60005b83811015620004f7578082015181840152602081019050620004da565b8381111562000507576000848401525b50505050565b6000620005246200051e84620004a1565b62000482565b90508281526020810184848401111562000543576200054262000407565b5b62000550848285620004d7565b509392505050565b600082601f83011262000570576200056f62000402565b5b8151620005828482602086016200050d565b91505092915050565b60008060408385031215620005a557620005a462000393565b5b6000620005b585828601620003eb565b925050602083015167ffffffffffffffff811115620005d957620005d862000398565b5b620005e78582860162000558565b9150509250929050565b600082825260208201905092915050565b7f455243313936373a206e657720696d706c656d656e746174696f6e206973206e60008201527f6f74206120636f6e747261637400000000000000000000000000000000000000602082015250565b600062000660602d83620005f1565b91506200066d8262000602565b604082019050919050565b60006020820190508181036000830152620006938162000651565b9050919050565b600081519050919050565b600081905092915050565b6000620006bd826200069a565b620006c98185620006a5565b9350620006db818560208601620004d7565b80840191505092915050565b6000620006f58284620006b0565b915081905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b600062000738601d83620005f1565b9150620007458262000700565b602082019050919050565b600060208201905081810360008301526200076b8162000729565b9050919050565b600081519050919050565b60006200078a8262000772565b620007968185620005f1565b9350620007a8818560208601620004d7565b620007b3816200040c565b840191505092915050565b60006020820190508181036000830152620007da81846200077d565b905092915050565b61044d80620007f26000396000f3fe6080604052366100135761001161001d565b005b61001b61001d565b005b610025610091565b610035610030610093565b6100a2565b565b606061005c83836040518060600160405280602781526020016103f1602791396100c8565b905092915050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b565b600061009d61014e565b905090565b3660008037600080366000845af43d6000803e80600081146100c3573d6000f35b3d6000fd5b60606000808573ffffffffffffffffffffffffffffffffffffffff16856040516100f291906102e5565b600060405180830381855af49150503d806000811461012d576040519150601f19603f3d011682016040523d82523d6000602084013e610132565b606091505b5091509150610143868383876101a5565b925050509392505050565b600061017c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b610087565b60000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b6060831561020857600083511415610200576101c085610064565b6101ff576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016101f690610359565b60405180910390fd5b5b829050610213565b610212838361021b565b5b949350505050565b60008251111561022e5781518083602001fd5b806040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161026291906103ce565b60405180910390fd5b600081519050919050565b600081905092915050565b60005b8381101561029f578082015181840152602081019050610284565b838111156102ae576000848401525b50505050565b60006102bf8261026b565b6102c98185610276565b93506102d9818560208601610281565b80840191505092915050565b60006102f182846102b4565b915081905092915050565b600082825260208201905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b6000610343601d836102fc565b915061034e8261030d565b602082019050919050565b6000602082019050818103600083015261037281610336565b9050919050565b600081519050919050565b6000601f19601f8301169050919050565b60006103a082610379565b6103aa81856102fc565b93506103ba818560208601610281565b6103c381610384565b840191505092915050565b600060208201905081810360008301526103e88184610395565b90509291505056fe416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c6564a26469706673582212203d3a4fa7bfbe885990cb0eecab0fe4880b9db576636d7b89055e37abf244b25b64736f6c634300080c0033416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c65640000000000000000000000005c9eb5d6a6c2c1b3efc52255c0b356f116f6f66d00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000024c4d66de8000000000000000000000000a5a1dddef094095afb7b6e322de72961df2e198800000000000000000000000000000000000000000000000000000000
#include <solc/solc.hpp>
Data getProxyContractInitCode(implementation_address, init_bytecode) {
    // The Solidity source code of the ERC1967Proxy smart contract
    std::string contract_source = R"(
        copy paste from https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/ERC1967/ERC1967Proxy.sol
    )";

    // Compile the contract source code
    std::string contract_name = "ERC1967Proxy";
    solc::Compiler compiler;
    solc::Compiler::Result result = compiler.compile(contract_source); // 0x608060405260405162000c5138038062000c51833981810160405281019062000029919062000580565b6200003d828260006200004560201b60201c565b5050620007d7565b62000056836200008860201b60201c565b600082511180620000645750805b156200008357620000818383620000df60201b620000371760201c565b505b505050565b62000099816200011560201b60201c565b8073ffffffffffffffffffffffffffffffffffffffff167fbc7cd75a20ee27fd9adebab32041f755214dbc6bffa90cc0225b39da2e5c2d3b60405160405180910390a250565b60606200010d838360405180606001604052806027815260200162000c2a60279139620001eb60201b60201c565b905092915050565b6200012b816200027d60201b620000641760201c565b6200016d576040517f08c379a000000000000000000000000000000000000000000000000000000000815260040162000164906200066d565b60405180910390fd5b80620001a77f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b620002a060201b620000871760201c565b60000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b60606000808573ffffffffffffffffffffffffffffffffffffffff1685604051620002179190620006dc565b600060405180830381855af49150503d806000811462000254576040519150601f19603f3d011682016040523d82523d6000602084013e62000259565b606091505b50915091506200027286838387620002aa60201b60201c565b925050509392505050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b606083156200031a5760008351036200031157620002ce856200027d60201b60201c565b62000310576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401620003079062000745565b60405180910390fd5b5b8290506200032d565b6200032c83836200033560201b60201c565b5b949350505050565b600082511115620003495781518083602001fd5b806040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016200037f9190620007b3565b60405180910390fd5b6000604051905090565b600080fd5b600080fd5b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000620003c9826200039c565b9050919050565b620003db81620003bc565b8114620003e757600080fd5b50565b600081519050620003fb81620003d0565b92915050565b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b62000456826200040b565b810181811067ffffffffffffffff821117156200047857620004776200041c565b5b80604052505050565b60006200048d62000388565b90506200049b82826200044b565b919050565b600067ffffffffffffffff821115620004be57620004bd6200041c565b5b620004c9826200040b565b9050602081019050919050565b60005b83811015620004f6578082015181840152602081019050620004d9565b60008484015250505050565b6000620005196200051384620004a0565b62000481565b90508281526020810184848401111562000538576200053762000406565b5b62000545848285620004d6565b509392505050565b600082601f83011262000565576200056462000401565b5b81516200057784826020860162000502565b91505092915050565b600080604083850312156200059a576200059962000392565b5b6000620005aa85828601620003ea565b925050602083015167ffffffffffffffff811115620005ce57620005cd62000397565b5b620005dc858286016200054d565b9150509250929050565b600082825260208201905092915050565b7f455243313936373a206e657720696d706c656d656e746174696f6e206973206e60008201527f6f74206120636f6e747261637400000000000000000000000000000000000000602082015250565b600062000655602d83620005e6565b91506200066282620005f7565b604082019050919050565b60006020820190508181036000830152620006888162000646565b9050919050565b600081519050919050565b600081905092915050565b6000620006b2826200068f565b620006be81856200069a565b9350620006d0818560208601620004d6565b80840191505092915050565b6000620006ea8284620006a5565b915081905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b60006200072d601d83620005e6565b91506200073a82620006f5565b602082019050919050565b6000602082019050818103600083015262000760816200071e565b9050919050565b600081519050919050565b60006200077f8262000767565b6200078b8185620005e6565b93506200079d818560208601620004d6565b620007a8816200040b565b840191505092915050565b60006020820190508181036000830152620007cf818462000772565b905092915050565b61044380620007e76000396000f3fe6080604052366100135761001161001d565b005b61001b61001d565b005b610025610091565b610035610030610093565b6100a2565b565b606061005c83836040518060600160405280602781526020016103e7602791396100c8565b905092915050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b565b600061009d61014e565b905090565b3660008037600080366000845af43d6000803e80600081146100c3573d6000f35b3d6000fd5b60606000808573ffffffffffffffffffffffffffffffffffffffff16856040516100f291906102db565b600060405180830381855af49150503d806000811461012d576040519150601f19603f3d011682016040523d82523d6000602084013e610132565b606091505b5091509150610143868383876101a5565b925050509392505050565b600061017c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b610087565b60000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b606083156102075760008351036101ff576101bf85610064565b6101fe576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016101f59061034f565b60405180910390fd5b5b829050610212565b610211838361021a565b5b949350505050565b60008251111561022d5781518083602001fd5b806040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161026191906103c4565b60405180910390fd5b600081519050919050565b600081905092915050565b60005b8381101561029e578082015181840152602081019050610283565b60008484015250505050565b60006102b58261026a565b6102bf8185610275565b93506102cf818560208601610280565b80840191505092915050565b60006102e782846102aa565b915081905092915050565b600082825260208201905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b6000610339601d836102f2565b915061034482610303565b602082019050919050565b600060208201905081810360008301526103688161032c565b9050919050565b600081519050919050565b6000601f19601f8301169050919050565b60006103968261036f565b6103a081856102f2565b93506103b0818560208601610280565b6103b98161037a565b840191505092915050565b600060208201905081810360008301526103de818461038b565b90509291505056fe416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c6564a2646970667358221220e57dd3eafc9985be746025b6d82d4f011b9a7bb3db56f9a1eb7eadfddd376b6064736f6c63430008110033416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c6564

    // Get the bytecode for the specified contract
    std::string bytecode = result.contracts.at(contract_name).bytecode;

    // Encode the constructor arguments
    bytecode += solc::encode_hex(implementation_address);
    bytecode += solc::encode_hex(init_bytecode);

    // The bytecode can be used to deploy the contract
    return bytecode;
}

// Test case:
// from: 0xd9145CCE52D386f254917e481eB44e9943F39138
// salt: 0
// initCodeHash: 0x013ef71ce2c00bc40f21a82848eb9d15b22d1b531df07dbd873a95534058942e
// returns: 0xf6548aAD02E0B355a0778ACc7799FD83B2506b58
Data create2Address(const std::string& from, const Data& salt, const Data& initCodeHash) {
    if (salt.size() != 32) {
        throw std::runtime_error("Error: salt must be 32 bytes.");
    }
    if (initCodeHash.size() != 32) {
        throw std::runtime_error("Error: initCodeHash must be 32 bytes.");
    }
    Data input = {0xff};
    append(input, parse_hex(from));
    append(input, salt);
    append(input, initCodeHash);
    auto hash = Hash::keccak256(input);
    return Data(hash.end() - 20, hash.end());
}
```
**Describe alternatives you've considered**

There are currently two working alternatives:
- The official ERC-4337 entry point smart contract which has a function to calculate user address. This isn't viable because in order to use that we need to send a request to a RPC, while we would prefer to be able to calculate user address locally, the same way as for EOAs.
- Ethers.js has the same functionality, however it isn't available to use from both wallet-core and TW mobile apps.

### PR DESCRIPTION
**Pull request by Milerius**
## Description

fix #2869 

## How to test

run C++ unit tests, check value from: https://eips.ethereum.org/EIPS/eip-1014

### CODE REVIEW

File: include/TrustWalletCore/TWEthereumEip1014.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+struct TWEthereumEip1014;
+
+/// EIP-1014: Skinny CREATE2 (guess smart contract create2 address)
+///
+/// \param fromEthAddress valid eth address
+/// \param salt always 32 bytes stack item
+/// \param initCodeHash The init_code is the code that, when executed, produces the runtime bytecode that will be placed into the state, and which typically is used by high level languages to implement a ‘constructor’. Need to be provided hashed with keccak256
+/// \return Ethereum resulting address
+TW_EXPORT_STATIC_METHOD
+TWString* _Nonnull TWEthereumEip1014AddressCreate2(TWString* _Nonnull fromEthAddress, TWData* _Nonnull salt, TWData* _Nonnull initCodeHash);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWEthereumEip1014.h:25]: Can you consolidate all Ethereum related methods into one interface? This looks badly organized and need to update docs every time for new added interface

**Review by catenocrypt — APPROVED** [review]

File: tests/chains/Ethereum/EIP1014Tests.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    TEST(EthereumEip1014, Example6) {
+        const std::string& from = "0x0000000000000000000000000000000000000000";
+        const Data salt = parse_hex("0x0000000000000000000000000000000000000000000000000000000000000000");
+        Data initCodeHash = Hash::keccak256(parse_hex("0x"));
+        const auto& addressData = Ethereum::create2Address(from, salt, initCodeHash);
+        ASSERT_EQ(Ethereum::checksumed(Ethereum::Address(hexEncoded(addressData))), "0xE33C0C7F7df4809055C3ebA6c09CFe4BaF1BD9e0");
+        ASSERT_EQ(Ethereum::create2AddressString(from, salt, initCodeHash), "0xE33C0C7F7df4809055C3ebA6c09CFe4BaF1BD9e0");
+    }
```
Comment (rsrbk) [inline tests/chains/Ethereum/EIP1014Tests.cpp:103]:
```suggestion
    }
    
    TEST(EthereumEip1014, Example7) {
        const std::string& from = "0x7EF2e0048f5bAeDe046f6BF797943daF4ED8CB47";
        const Data salt = parse_hex("0x0000000000000000000000000000000000000000000000000000000000000000");
        Data initCodeHash = Hash::keccak256(parse_hex("0x608060405260405162000c5138038062000c51833981810160405281019062000029919062000580565b6200003d828260006200004560201b60201c565b5050620007d7565b62000056836200008860201b60201c565b600082511180620000645750805b156200008357620000818383620000df60201b620000371760201c565b505b505050565b62000099816200011560201b60201c565b8073ffffffffffffffffffffffffffffffffffffffff167fbc7cd75a20ee27fd9adebab32041f755214dbc6bffa90cc0225b39da2e5c2d3b60405160405180910390a250565b60606200010d838360405180606001604052806027815260200162000c2a60279139620001eb60201b60201c565b905092915050565b6200012b816200027d60201b620000641760201c565b6200016d576040517f08c379a000000000000000000000000000000000000000000000000000000000815260040162000164906200066d565b60405180910390fd5b80620001a77f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b620002a060201b620000871760201c565b60000160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050565b60606000808573ffffffffffffffffffffffffffffffffffffffff1685604051620002179190620006dc565b600060405180830381855af49150503d806000811462000254576040519150601f19603f3d011682016040523d82523d6000602084013e62000259565b606091505b50915091506200027286838387620002aa60201b60201c565b925050509392505050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b606083156200031a5760008351036200031157620002ce856200027d60201b60201c565b62000310576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401620003079062000745565b60405180910390fd5b5b8290506200032d565b6200032c83836200033560201b60201c565b5b949350505050565b600082511115620003495781518083602001fd5b806040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016200037f9190620007b3565b60405180910390fd5b6000604051905090565b600080fd5b600080fd5b600073ffffffffffffffffffffffffffffffffffffffff82169050919050565b6000620003c9826200039c565b9050919050565b620003db81620003bc565b8114620003e757600080fd5b50565b600081519050620003fb81620003d0565b92915050565b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b62000456826200040b565b810181811067ffffffffffffffff821117156200047857620004776200041c565b5b80604052505050565b60006200048d62000388565b90506200049b82826200044b565b919050565b600067ffffffffffffffff821115620004be57620004bd6200041c565b5b620004c9826200040b565b9050602081019050919050565b60005b83811015620004f6578082015181840152602081019050620004d9565b60008484015250505050565b6000620005196200051384620004a0565b62000481565b90508281526020810184848401111562000538576200053762000406565b5b62000545848285620004d6565b509392505050565b600082601f83011262000565576200056462000401565b5b81516200057784826020860162000502565b91505092915050565b600080604083850312156200059a576200059962000392565b5b6000620005aa85828601620003ea565b925050602083015167ffffffffffffffff811115620005ce57620005cd62000397565b5b620005dc858286016200054d565b9150509250929050565b600082825260208201905092915050565b7f455243313936373a206e657720696d706c656d656e746174696f6e206973206e60008201527f6f74206120636f6e747261637400000000000000000000000000000000000000602082015250565b600062000655602d83620005e6565b91506200066282620005f7565b604082019050919050565b60006020820190508181036000830152620006888162000646565b9050919050565b600081519050919050565b600081905092915050565b6000620006b2826200068f565b620006be81856200069a565b9350620006d0818560208601620004d6565b80840191505092915050565b6000620006ea8284620006a5565b915081905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b60006200072d601d83620005e6565b91506200073a82620006f5565b602082019050919050565b6000602082019050818103600083015262000760816200071e565b9050919050565b600081519050919050565b60006200077f8262000767565b6200078b8185620005e6565b93506200079d818560208601620004d6565b620007a8816200040b565b840191505092915050565b60006020820190508181036000830152620007cf818462000772565b905092915050565b61044380620007e76000396000f3fe6080604052366100135761001161001d565b005b61001b61001d565b005b610025610091565b610035610030610093565b6100a2565b565b606061005c83836040518060600160405280602781526020016103e7602791396100c8565b905092915050565b6000808273ffffffffffffffffffffffffffffffffffffffff163b119050919050565b6000819050919050565b565b600061009d61014e565b905090565b3660008037600080366000845af43d6000803e80600081146100c3573d6000f35b3d6000fd5b60606000808573ffffffffffffffffffffffffffffffffffffffff16856040516100f291906102db565b600060405180830381855af49150503d806000811461012d576040519150601f19603f3d011682016040523d82523d6000602084013e610132565b606091505b5091509150610143868383876101a5565b925050509392505050565b600061017c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc60001b610087565b60000160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b606083156102075760008351036101ff576101bf85610064565b6101fe576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016101f59061034f565b60405180910390fd5b5b829050610212565b610211838361021a565b5b949350505050565b60008251111561022d5781518083602001fd5b806040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161026191906103c4565b60405180910390fd5b600081519050919050565b600081905092915050565b60005b8381101561029e578082015181840152602081019050610283565b60008484015250505050565b60006102b58261026a565b6102bf8185610275565b93506102cf818560208601610280565b80840191505092915050565b60006102e782846102aa565b915081905092915050565b600082825260208201905092915050565b7f416464726573733a2063616c6c20746f206e6f6e2d636f6e7472616374000000600082015250565b6000610339601d836102f2565b915061034482610303565b602082019050919050565b600060208201905081810360008301526103688161032c565b9050919050565b600081519050919050565b6000601f19601f8301169050919050565b60006103968261036f565b6103a081856102f2565b93506103b0818560208601610280565b6103b98161037a565b840191505092915050565b600060208201905081810360008301526103de818461038b565b90509291505056fe416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c6564a2646970667358221220e57dd3eafc9985be746025b6d82d4f011b9a7bb3db56f9a1eb7eadfddd376b6064736f6c63430008110033416464726573733a206c6f772d6c6576656c2064656c65676174652063616c6c206661696c6564000000000000000000000000d9ec9e840bb5df076dbbb488d01485058f421e5800000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000024c4d66de8000000000000000000000000be8fa0112dcb7d21dc63645b633073651e19934800000000000000000000000000000000000000000000000000000000"));
        const auto& addressData = Ethereum::create2Address(from, salt, initCodeHash);
        ASSERT_EQ(Ethereum::checksumed(Ethereum::Address(hexEncoded(addressData))), "0x4455e5f0038795939c001aa4d296A45956C460AA");
        ASSERT_EQ(Ethereum::create2AddressString(from, salt, initCodeHash), "0x4455e5f0038795939c001aa4d296A45956C460AA");
    }
```

Comment (rsrbk) [reply tests/chains/Ethereum/EIP1014Tests.cpp:103]: Adding one more real example for a ERC-4337 SCW

**Review by rsrbk — APPROVED** [review]

File: src/Ethereum/EIP1014.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        throw std::runtime_error("Error: salt must be 32 bytes.");
+    }
+    if (initCodeHash.size() != 32) {
+        throw std::runtime_error("Error: initCodeHash must be 32 bytes.");
+    }
+    Data input = {0xff};
+    append(input, parse_hex(from));
+    append(input, salt);
+    append(input, initCodeHash);
+    auto hash = Hash::keccak256(input);
```
Comment (rsrbk) [inline src/Ethereum/EIP1014.cpp:25]: Shouldn't we keccak the input before passing it into this function? (to follow the same implementation as ethers.js)

Comment (Milerius) [reply src/Ethereum/EIP1014.cpp:25]: Only initCodeHash need to be hashed before calling the function - ether.js hash also inside the concatenation of everything.

**Review by hewigovens — APPROVED** [review]

**Pull request merged**